src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,428 @@
+            self.modulus()
+        )
+
+    def __str__(self):
+        return "fmpz_mod_ctx({})".format(
+            self.modulus()
+        )
+
+    def __call__(self, val):
+        if not typecheck(val, fmpz):
+            val = any_as_fmpz(val)
+            if val is NotImplemented:
+                raise NotImplementedError("TODO")
+
+        return fmpz_mod(val, self)
+
+
+cdef class fmpz_mod(flint_scalar):
+    def __init__(self, val, ctx):
+        self.ctx = ctx
+
+        if not typecheck(val, fmpz):
+            val = any_as_fmpz(val)
+            if val is NotImplemented:
+                raise NotImplementedError("TODO")
+
+        fmpz_mod_set_fmpz(self.val, (<fmpz>val).val, self.ctx.val)
+
+    def is_zero(self):
+        return self == 0
+    
+    # TODO: kind of pointless, as we always ensure canonical on init?
+    def is_canonical(self):
+        cdef bint res
+        res = fmpz_mod_is_canonical(self.val, self.ctx.val)
+        return res == 1
+
+    def is_one(self):
+        cdef bint res
+        res = fmpz_mod_is_one(self.val, self.ctx.val)
+        return res == 1
+
+    def __richcmp__(s, t, int op):
+        cdef bint res
+        if op != 2 and op != 3:
+            raise TypeError("fmpz_mod cannot be ordered")
+        # TODO: is this the best method for comparison?
+        if typecheck(s, fmpz_mod) and typecheck(t, fmpz_mod):
+            res = ((<fmpz_mod>s).val[0] == (<fmpz_mod>t).val[0]) and \
+                  ((<fmpz_mod>s).ctx.val.n == (<fmpz_mod>t).ctx.val.n)
+            if op == 2:
+                return res
+            else:
+                return not res
+        # TODO: is this the best method for comparison?
+        # Seems like I'm doing too many type conversions?
+        elif typecheck(s, fmpz_mod) and typecheck(t, int):
+            res = int(s) == t % (<fmpz_mod>s).ctx.modulus()
+            if op == 2:
+                return res
+            else:
+                return not res
+        return NotImplemented
+
+    def __bool__(self):
+        return not self.is_zero()
+
+    def __repr__(self):
+        return "fmpz_mod({}, {})".format(
+            fmpz_get_intlong(self.val),
+            self.ctx.modulus()
+        )
+
+    # TODO: seems ugly...
+    def __hash__(self):
+        return  hash(
+            (int(self), int(self.ctx.modulus()))
+        )
+
+    def __int__(self):
+        return fmpz_get_intlong(self.val)
+
+    def __str__(self):
+        return str(int(self))
+
+    # ---------------- #
+    #    Arithmetic    #
+    # ---------------- #
+
+    def __pos__(self):
+        return self
+
+    def __neg__(self):
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
+
+        fmpz_mod_neg(
+            res.val, self.val, 
+            (<fmpz_mod_ctx_t>self.ctx.val))
+        return res
+
+    def __add__(self, other):
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
```

Comment:
There are multiple exits from this function that do not return `res`. I think that `__new__` should not be called until we are sure that `res` will be returned.

The call to `__new__` is equivalent to calling `fmpz_init`

Reply:
> The call to `__new__` is equivalent to calling `fmpz_init`

Actually this is not true but it *should* be. We need a `__cinit__` method like for `fmpz`.

Reply:
> The call to `__new__` is equivalent to calling `fmpz_init`

The point here is that any call to `fmpz_init` should eventually be paired with a call to `fmpz_clear`. That can happen here if we are not going to return `res`. Otherwise it should happen in dealloc.

Reply:
So I understand, we want to have both a 

```cython
def __cinit__(self):
    fmpz_init(self.val)
    
def __dealloc__(self):
    fmpz_clear(self.val)
```

Included in the `fmpz` class, something similar for `fmpz_mod` and aslo I need to ensure that I am only calling `__new__` in places when I know `res` is returned?

Reply:
> I need them on the `fmpz_mod()` class?

Yes, and also `fmpz_mod_ctx`.

Reply:
There is no `fmpz_mod_init` :/

https://flintlib.org/doc/fmpz_mod.html

Reply:
> There is no `fmpz_mod_init`

That's because it is `fmpz_init`.